### PR TITLE
CQL Search

### LINKED ISSUE #5: Handling start_datetime and end_datetime?
There's some datasets that do not have a datetime, but instead have a start_datetime and end_datetime. For example, a landcover dataset that represents a year of data - there's no single datetime to represent the data. STAC specifies that in this case `datetime` is `null` and the start_datetime and end_datetime is set.

How can pgstac handle these types of items?

### LINKED ISSUE #14: Support CQL JSON for Queries

### LINKED ISSUE #15: Move search to use reusable cursor code for scanning through partitions.

### LINKED ISSUE #35: Remove timeStamp from search response
https://github.com/stac-utils/pgstac/blob/main/sql/004_search.sql#L561

This field doesn't match the STAC API response and is extraneous; I think we should remove to avoid confusion and stick to standard fields.

### LINKED ISSUE #36: Add support for the Context Extension

### PR DESCRIPTION
**Pull request by bitner**
Refactors search to use CQL JSON filter spec for search (Fixes #14). Maintains backwards compatibility with STAC query json ("filter" uses CQL, items passed into a search using "query" are remapped into CQL and added to anything in the "filter"). 

This PR involves a reworking of data layout for items and as such ends up moving all existing data. Migrations from large earlier versions will take a while!

Reasons for the data layout changes:
 - previously the index we had on properties was limited to lowercase (all text comparisons using the query spec were lower case, CQL is case sensitive
 - allows indexed equality lookup of properties to an arbitrary depth
 - Use triggers to update instantiated fields rather than generated columns
 - date queries take into account datetime and start_datetime/end_datetime (fixes #5)
 - instantiated columns have slightly more creation overhead than generated columns but the query planner makes much better choices and performance is drastically improved

The partition-wise scan process was generalized to use cursors and to better leverage some of the query planner. (Fixes #15).

Splits scripts to organize things a bit more.

Add many PGTap tests.

Adds support for Context Extension (Fixes #36, #37)
 - Running a full count for a query can still take considerably longer than just returning a limited set of rows. PGStac will estimate the row count using PostgreSQL Explain which leverages the query planner and will do a full count only when the estimated count is less than 10000. Passing "context":true in the search will force PGStac to do a full count even when the estimate is large. Both estimated count and full count are cached in the searches table and are cached for 1 day so subsequent identical queries will not have to do the full count.

Removes the timeStamp from the search response (Fixes #35).

Bumps version to 0.3.0.

### CODE REVIEW

**Review by lossyrob — APPROVED** [review]
Looks great! Some minor comments/questions inline.

File: test/pgtap/003_items.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
+SELECT has_table('pgstac'::name, 'items'::name);
+
+-- SELECT is_indexed('pgstac'::name, 'items'::name, ARRAY['datetime','id']);
+-- SELECT is_indexed('pgstac'::name, 'items'::name, 'properties');
+SELECT is_indexed('pgstac'::name, 'items'::name, 'geometry');
+SELECT is_indexed('pgstac'::name, 'items'::name, 'collection_id');
+
+-- SELECT has_type('pgstac'::name, 'item'::name);
+SELECT is_partitioned('pgstac'::name,'items'::name);
```
[the remark this answers is not in the document]
Comment (bitner) [reply test/pgtap/003_items.sql:9]: We are still using partitions, just not using pg_partman to manage them.

File: scripts/stageversion [hunk 2]
```diff
@@ -16,8 +16,8 @@ Create a new tag
 
 if [ "${BASH_SOURCE[0]}" = "${0}" ]; then
     echo "Updating Version..."
-    DIRTY=$(git status --porcelain=v1 2>/dev/null | wc -l)
-    [ $DIRTY -gt 0 ] && (echo "Git has dirty files, stash or commit before staging version."; exit 1)
+    # DIRTY=$(git status --porcelain=v1 2>/dev/null | wc -l)
+    # [ $DIRTY -gt 0 ] && (echo "Git has dirty files, stash or commit before staging version."; exit 1)
```
[the remark this answers is not in the document]
Comment (bitner) [reply scripts/stageversion]: Yes. Lines removed in latest commit.

File: test/pgtap/001_core.sql [hunk 3]
```diff
@@ -0,0 +1,19 @@
+-- Check that schema exists
+SELECT has_schema('pgstac'::name);
+
+-- Check that PostGIS and PG_Partman extensions are installed and available on the path
+SELECT has_extension('postgis');
+SELECT has_extension('pg_partman');
```
[the remark this answers is not in the document]
Comment (bitner) [reply test/pgtap/001_core.sql]: Good catch. Yes.

File: test/pgtap/003_items.sql [hunk 4]
```diff
@@ -0,0 +1,31 @@
+SELECT has_table('pgstac'::name, 'items'::name);
+
+-- SELECT is_indexed('pgstac'::name, 'items'::name, ARRAY['datetime','id']);
```
[the remark this answers is not in the document]
Comment (bitner) [reply test/pgtap/003_items.sql:3]: Good catch. I had commented some indexes while testing migrations. Checked through to make sure they were all put back in last commit.

**Pull request merged**